### PR TITLE
Add tutorial on creating a source plugin

### PR DESCRIPTION
**Pull request by m-allanson**
The plugin is based on @jlengstorf's [`gatsby-source-pixabay`](https://github.com/jlengstorf/gatsby-source-pixabay) plugin.

I've written it using Node 6 compatible JavaScript to avoid worrying about babel.

All feedback welcome!

Edit: [view formatted version](https://github.com/m-allanson/gatsby/blob/ce997a6d8de87613791299fffbab3fe39fc4c0a2/docs/docs/source-plugin-tutorial.md)

### CODE REVIEW

File: docs/docs/source-plugin-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,391 @@
+---
+title: "Source plugin tutorial"
+---
+
+Creating your own source plugin.
+
+# What this tutorial covers
```
Comment (KyleAMathews) [inline docs/docs/source-plugin-tutorial.md]: These should all be h2s

File: docs/docs/source-plugin-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+In this tutorial you'll create your own source plugin. Your plugin will source data from [pixabay.com](https://pixabay.com) allowing you to add Pixabay images to any Gatsby site.
+
+# What is a source plugin?
+
+Source plugins "source" data from remote or local locations into what Gatsby calls [nodes](https://www.gatsbyjs.org/docs/node-interface/).
```
Comment (KyleAMathews) [inline docs/docs/source-plugin-tutorial.md]: use relative links

File: docs/docs/source-plugin-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    }
+  }
+}
+```
+
+Or [open the query from this link](<http://localhost:8000/___graphql?query=%7B%0A%20%20allPixabayPhoto(limit%3A%2010)%20%7B%0A%20%20%20%20edges%20%7B%0A%20%20%20%20%20%20node%20%7B%0A%20%20%20%20%20%20%20%20largeImageURL%0A%20%20%20%20%20%20%20%20pageURL%0A%20%20%20%20%20%20%20%20tags%0A%20%20%20%20%20%20%20%20user%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A>).
+
+Experiment with different options in your `gatsby-config.js` file to see how that affects your query results. The [Pixabay API docs](https://pixabay.com/api/docs/#api_search_images) might be a useful reference.
+
+### Publishing a plugin
```
Comment (jlengstorf) [inline docs/docs/source-plugin-tutorial.md]: This is a new section, so I'd probably go back to H2 here.

File: docs/docs/source-plugin-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+
+You've built a local plugin for your project, but what if you want to share it with others? `npm` allows you to publish your plugins, check out the npm docs on [How to Publish & Update a Package](https://docs.npmjs.com/getting-started/publishing-npm-packages) for more info.
+
+> Don't forget to edit your plugin's `package.json` file to include info about your plugin.
+
+### Summary
```
Comment (jlengstorf) [inline docs/docs/source-plugin-tutorial.md]: H2

File: docs/docs/source-plugin-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+You've written a local Gatsby plugin that:
+
+* can be configured with an entry in your `gatsby-config.js` file
+* requests data from a third-party API
+* pulls the API data into Gatsby's node system
+* allows the data to be queried with GraphQL
+
+Congratulations!
+
+### Where next?
```
Comment (jlengstorf) [inline docs/docs/source-plugin-tutorial.md]: H2

File: docs/docs/source-plugin-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+Your plugin has been adapted from Jason Lengstorf's [gatsby-source-pixabay plugin](https://www.npmjs.com/package/gatsby-source-pixabay). Check out [the source on GitHub](https://github.com/jlengstorf/gatsby-source-pixabay).
+
+Try adding new features to your plugin, for example - download images from Pixabay, improve error handling, add documentation or automated tests.
+
+Check out Gatsby's docs on [plugin authoring](/docs/plugin-authoring/) and [creating a source plugin](/docs/create-source-plugin).
+
+#### A note on JavaScript versions
```
Comment (jlengstorf) [inline docs/docs/source-plugin-tutorial.md]: H3

File: docs/docs/source-plugin-tutorial.md [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+
+> **NOTE:** You'll need a Pixabay API key which you can get by [registering for a Pixabay account](https://pixabay.com/en/accounts/register/). Your API key is in the [“Search Images” section of the Pixabay API docs](https://pixabay.com/api/docs/#api_search_images).
+
+### An example API request
+
+Pixabay's [API documentation](https://pixabay.com/api/docs/#api_search_images) describes how their API works. Here's an example that uses a few options to search for photos:
+
+`https://pixabay.com/api/?q=yellow+flowers&editors_choice=true&pretty=true&key=<YOUR_API_KEY_HERE>`
+
+Take the above URL and paste it in to a browser to see Pixabay's response to your query. It's given us a list of photos matching the query "yellow flowers" that have received an Editor's Choice award.
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: *It gives you

File: docs/docs/source-plugin-tutorial.md [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+    "node-fetch": "^2.1.2",
+    "query-string": "^6.0.0"
+  }
+```
+
+With the setup done, move on to adding the plugin's functionality.
+
+### Create a `gatsby-node.js` file
+
+Create a new file called `gatsby-node.js`, and add the following:
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: at the root of the project directory? or where?

File: docs/docs/source-plugin-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+
+  // Gatsby adds a configOption that's not needed for this plugin, delete it
+  delete configOptions.plugins;
+
+  // plugin code goes here...
+  console.log("Testing my plugin", configOptions);
+};
+```
+
+Let's break this down, you start by importing the dependencies that you added earlier (along with one built in dependency):
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: *What did you just do by adding this code? You started by...

File: docs/docs/source-plugin-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+
+```js
+const crypto = require("crypto");
+const fetch = require("node-fetch");
+const queryString = require("query-string");
+```
+
+Then you implement Gatsby's [`sourceNodes` API](/docs/node-apis/#sourceNodes) which Gatsby will run as part of its bootstrap process. When Gatsby calls `sourceNodes`, it'll pass in some helper functions (`boundActionCreators` and `createNodeId`) along with any config options that are provided in your project's `gatsby-config.js` file:
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: *implemented

File: docs/docs/source-plugin-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+        // For each query result (or 'hit')
+        data.hits.forEach(photo => {
+          console.log("Photo data is:", photo)
+        })
+      })
+  )
+}
+```
+
+You've added code that fetches photo data from the Pixabay API. For now your plugin logs that data but doesn't do anything else. Check that you can see the logged photo data by restarting `gatsby dev`, this time you should see a series of results like:
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: *For now,

Comment (shannonbux) [reply docs/docs/source-plugin-tutorial.md]: *`gatsby dev`. This time...

File: docs/docs/source-plugin-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+          createNode(nodeData)
+        })
+      })
+  )
+}
+```
+
+### Query for results
+
+Your plugin is ready. Restart `gatsby dev` and open a browser at [http://localhost:8000/\_\_\_graphql](http://localhost:8000/___graphql). The Pixabay data can be queried from here, try:
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: *from here. Try...

File: docs/docs/source-plugin-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+}
+```
+
+Or [open the query from this link](<http://localhost:8000/___graphql?query=%7B%0A%20%20allPixabayPhoto(limit%3A%2010)%20%7B%0A%20%20%20%20edges%20%7B%0A%20%20%20%20%20%20node%20%7B%0A%20%20%20%20%20%20%20%20largeImageURL%0A%20%20%20%20%20%20%20%20pageURL%0A%20%20%20%20%20%20%20%20tags%0A%20%20%20%20%20%20%20%20user%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A>).
+
+Experiment with different options in your `gatsby-config.js` file to see how that affects your query results. The [Pixabay API docs](https://pixabay.com/api/docs/#api_search_images) might be a useful reference.
+
+## Publishing a plugin
+
+You've built a local plugin for your project, but what if you want to share it with others? `npm` allows you to publish your plugins, check out the npm docs on [How to Publish & Update a Package](https://docs.npmjs.com/getting-started/publishing-npm-packages) for more info.
```
Comment (shannonbux) [inline docs/docs/source-plugin-tutorial.md]: *plugins. Check out...

File: docs/docs/source-plugin-tutorial.md (hunk 8 above)
Comment (m-allanson) [reply docs/docs/source-plugin-tutorial.md]: In the same directory as the `package.json`, I'll clarify this with " in your `gatsby-source-pixabay` directory"

File: docs/docs/source-plugin-tutorial.md (hunk 9 above)
Comment (m-allanson) [reply docs/docs/source-plugin-tutorial.md]: 👍 I dropped _just_ from your edit, so it's "What did you do by adding this code? You started by.."

File: docs/docs/source-plugin-tutorial.md (hunk 13 above)
Comment (m-allanson) [reply docs/docs/source-plugin-tutorial.md]: Yeah I'm still a bit undecided about how much detail to go into here. The plugin authoring doc is mentioned later on so I skipped it here. And I wondered if mentioning the plugin library here might lead to a bunch of duplicate `gatsby-source-pixabay` plugins being added. 

What do you think? Maybe it's worth mentioning the plugin library in the _What's next?_ section? Although if someone is interested in writing their own plugin, can we assume they're already aware of the plugin library?

**Comment by m-allanson** [comment]
Thanks @shannonbux! It took a while to plan and write 😅. 

> perhaps the second half of the tutorial could use more headers to break it up.

Good idea, I've added one more header "Step by step through your `gatsby-node.js` file". Do you think that'll do it?

**Comment by m-allanson** [comment]
@KyleAMathews you might want this for the workshop too?

**Pull request merged**